### PR TITLE
W.I.P Add Getting Started guide for creating rules

### PR DESCRIPTION
**Pull request by fatbasstard**
Working on a "Getting Started" guide to explain the concept of Rules and how to build a rule.

Currently full of typo's and missing content, but building and tweaking it along the way. Please try it out and comment if you find stuff!

Todo:
- [x] Add code sample snippets
- [x] Remove all typo's 😉 
- [x] Get confirmation from a couple of people that it is correct

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/getting_started_rules.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+An overview of all the rules that are currently supported can be found [**here**](./rules.md).
+
+Besides the rules from `cfn-lint` itself, it can also process external rules (by using the `-a/--append-rules` argument), making it possible to integrate custom rulesets.
+
+### Rules of the rules
+When creating a rule for `cfn-lint`, keep the following rules in mind:
+
+* A rule has to be applicable to *ALL* users of the toolkit by default. If a rule contains specific business logic, create a custom rule.
+* A rule typically covers a single specific use case. It's not an exact science, but create multiple rules if you thing it's needed.
```
Comment (hoegertn) [inline docs/getting_started_rules.md]: [...] if you thin**k** [...]

Comment (MattHodge) [inline docs/getting_started_rules.md]: `thing` -> `think`

File: docs/getting_started_rules.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+* A rule is an `Error` (blocking) or `Warning` (not blocking). Create multiple specific rules if a rule can be both.
+* A rule should focus on it's own requirements. Missing a required property? That's handled by an [existing rule](/rules.md#E3003) already.
+
+These rules don't apply to custom rules sets, you're free to build your own rules in the way they suit you best.
+
+### Rule Matches
+A rule returns it findings by returning a list of `RuleMatch` object. A Match is an object containing information about the finding:
```
Comment (MattHodge) [inline docs/getting_started_rules.md]: Match -> `Match`

File: docs/getting_started_rules.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    def match(self, cfn):
+        """Basic Rule Matching"""
+
+        matches = list()
+
+        # Your Rule code goes here
+
+        return matches
+```
+Fill in the Meta data (`id`, `shortdesc`, `description`, `source_url`, `tags`) of your new rule, there are required attributed needed for processing the rule.
```
Comment (MattHodge) [inline docs/getting_started_rules.md]: `there are required attributed needed for processing the rule.` -> `these are required attributes needed for processing the rule`

File: docs/getting_started_rules.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+Since `cfn-lint` loads all rules from the `rules` folder as a plugin, the file and Class name must be the same.
+
+*At this point your new rule is already part of the linter's ruleset!*
+
+### Test your new Rule
+You can test your new rule by just running `cfn-lint`, but preferably your new Rule is validated by unit tests. In this way the working is guaranteed for the future, and different use-cases can be validated automatically.
+
+#### Fixtures
+Fixture templates are created under the [test/templates/good](/test/templates/good) (for templates with valid CloudFormation code) or [test/templates/bad](/test/templates/bad) (for template with "invalid" CloudFormation Code) folders.  
+Use `YAML` for the CLoudFormation templates for readability and the ability to add comments in the template
```
Comment (MattHodge) [inline docs/getting_started_rules.md]: `CLoudFormation` -> `CloudFormation`

**Comment by fatbasstard** [comment]
@hoegertn @MattHodge Thanx for the feedback already! Procssing now....

File: docs/getting_started_rules.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+      # Only check what we need. Other rules handle misconfigurations
+      if properties:
+          # Check the instance type specified
+          if properties.get('InstanceType') == 't2.small':
+
+            # The path is used to determine the line/column number when returning the Match information.
+            path = ['Resources', resource_name, 'Properties', 'InstanceType']
+            message = 'Please do not use t2.small'
+            matches.append(RuleMatch(path, message.format(resource_name)))
+```
```
Comment (SanderKnape) [inline docs/getting_started_rules.md]: Maybe you could add some links to already merged pull requests that add a single or multiple rules?

File: docs/getting_started_rules.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+* **py27**: Unit tests (Python 2.7)
+* **pylint27**: Python syntax check (Python 2.7)
+* **py36**: Unit tests (Python 3.6)
+* **pylint36**: Python syntax check (Python 3.6)
+
+## Code Snippets
+The skeleton code mention in the [Create new Rule](#create_new_rule) step contains no real code to keep it as simple as possible. This section contains a few simple snippets to get you started. For more complex solutions, take a deep-dive in the code of the [existing rules](/src/cfnlint/rules)!
+
+### Add a RuleMatch
+
```
Comment (SanderKnape) [inline docs/getting_started_rules.md]: Empty line can be removed.

File: docs/getting_started_rules.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+*At this point your new rule is already part of the linter's ruleset!*
+
+See the [code snippets](#code_snippets) section below for some examples of an implemention.
+
+### Test your new Rule
+You can test your new rule by just running `cfn-lint`, but preferably your new Rule is validated by unit tests. In this way the working is guaranteed for the future, and different use-cases can be validated automatically.
+
+#### Fixtures
+Fixture templates are created under the [test/templates/good](/test/templates/good) (for templates with valid CloudFormation code) or [test/templates/bad](/test/templates/bad) (for template with "invalid" CloudFormation Code) folders.
```
Comment (SanderKnape) [inline docs/getting_started_rules.md]: No need for quotes around "invalid" and you can remove capital letter for "Code"

File: docs/getting_started_rules.md (hunk 5 above)
Comment (fatbasstard) [reply docs/getting_started_rules.md]: That's stated in the introduction of the Snippets section.

Comment (mikecee) [reply docs/getting_started_rules.md]: Suggest an explicit reference to using/providing Fixtures at the end of the test case starting-point, something like

As you can see `test_file_negative()` in this unit test makes specific use of a CloudFormation template from the Fixtures folder.  It's important to provide examples of templates which _pass_ your new rule, and also templates which generate the expected _warning/error.

Comment (fatbasstard) [reply docs/getting_started_rules.md]: Goop point. Added

**Pull request merged**